typed-racket-lib/typed-racket/infer/infer-unit.rkt

```diff
@@ -790,37 +791,38 @@
                    ;; TODO figure out if there is a better subst here
                    [(? variance:inv) (i-subst null)]))))
      S))
-  (match (car (cset-maps C))
-    [(cons cmap (dmap dm))
-     (let* ([subst (hash-union
-                    (for/hash ([(k dc) (in-hash dm)])
-                      (define (c->t c) (constraint->type c (hash-ref idx-hash k variance:const)))
-                      (values
-                       k
-                       (match dc
-                         [(dcon fixed #f)
-                          (i-subst (map c->t fixed))]
-                         [(or (dcon fixed rest) (dcon-exact fixed rest))
-                          (i-subst/starred
-                           (map c->t fixed)
-                           (c->t rest))]
-                         [(dcon-dotted fixed dc dbound)
-                          (i-subst/dotted
-                           (map c->t fixed)
-                           (c->t dc)
-                           dbound)])))
-                    (for/hash ([(k v) (in-hash cmap)])
-                      (values k (t-subst (constraint->type v (hash-ref var-hash k variance:const))))))]
-            [subst (for/fold ([subst subst]) ([v (in-list X)])
-                     (let ([entry (hash-ref subst v #f)])
-                       ;; Make sure we got a subst entry for a type var
-                       ;; (i.e. just a type to substitute)
-                       ;; If we don't have one, there are no constraints on this variable
-                       (if (and entry (t-subst? entry))
-                           subst
-                           (hash-set subst v (t-subst Univ)))))])
-       ;; verify that we got all the important variables
-       (extend-idxs subst))]))
+  (for/list ([m (in-list (cset-maps C))])
```

Comment:
Can we thread `multiple-substitutions?` to here and avoid doing all this work when it's not needed?